piptools/repositories/pypi.py

```diff
@@ -284,10 +284,10 @@ def get_dependencies(self, ireq):
                         os.environ["PIP_REQ_TRACKER"] = prev_tracker
                     else:
                         del os.environ["PIP_REQ_TRACKER"]
-                try:
-                    self.wheel_cache.cleanup()
-                except AttributeError:
-                    pass
+
+                # WheelCache.cleanup() introduced in pip==10.0.0
```

Comment:
I don't see any need to change this. In fact, following [EAFP](https://docs.python.org/3/glossary.html#term-eafp) style is more Pythonic.
The only thing I'd change is using a CM instead of a try-block.
```suggestion
                # WheelCache.cleanup() introduced in pip==10.0.0
                with contextlib.suppress(AttributeError):
                    wheel_cache.cleanup()
```

Reply:
Thanks for the suggestion! This suppressor looks pythonic and way better than try-except. I like it.

A few notes about the explicit version checking:
1. the bug with unexpected AssertionError on `self` could be caught earlier
2. when we decide to remove the support of a specific pip version (e.g., pip=={version}), we would find the code by `git grep PIP_VERSION | grep {version}`.
3. if we'd be on the multi-threaded environment, the EAFP style would be better for this particular case. 
4. also, there is an excellent explanation originally posted by @nicoddemus:  https://twitter.com/nicoddemus/status/1146196274309414912. (yeah, he's talking about import statements, but it can be extrapolated)

I'm not against the suggested solution, but I'd prefer the uniform approach, which follows the current codebase.

Reply:
How is it related to `AssertionError`? Is it a typo for `AttributeError`?

Reply:
As for (2), grep would find it if you just add a proper comment which is necessary anyway to explain the intent behind the code.

Reply:
(3) not sure how it's related to multi-threading at all.

Reply:
(4) it may get quite complex with nested if-elses over time.

Reply:
> How is it related to `AssertionError`? Is it a typo for `AttributeError`?

Sorry, it's a typo, I meant the AttributeError.

> As for (2), grep would find it if you just add a proper comment which is necessary anyway to explain the intent behind the code.

Fair enough!

> (3) not sure how it's related to multi-threading at all.

The explanation is here https://docs.python.org/3/glossary.html#term-lbyl.

> (4) it may get quite complex with nested if-elses over time.

That's how we pay the cost of supporting pip>=8.1.1.


---

Also, `if-statement` makes type checkers happier.

Reply:
Hey @webknjaz! Gentle ping. If you feel strongly that we should use "EAFP style" here, so I'd rather accept your commit suggestion. Don't want to waste energy on the bikeshedding. Otherwise, let's merge it as is 😊

Reply:
Feel free to merge as-is.

Reply:
I'll mark this as resolved then.